gemonlinedb/src/common/GEMOnlineDBManagerWeb.cc

```diff
@@ -65,9 +152,87 @@ void gem::onlinedb::GEMOnlineDBManagerWeb::applicationPage(xgi::Input* in, xgi::
+  // first tab. Additional information is retrieved from other sources.
+
   CMSGEMOS_DEBUG("GEMOnlineDBManagerWeb::jsonUpdate");
   out->getHTTPResponseHeader().addHeader("Content-Type", "application/json");
-  *out << " { " << std::endl;
-  // fill JSON if needed
-  *out << " } " << std::endl;
+
+  json_auto_t *answer = janssonCheckPtr(json_object());
+
+  auto lock = ConfigurationManager::makeReadLock();
+  if (!lock.timed_lock(boost::posix_time::seconds(1))) {
+    // Error
+    janssonCheck(json_object_set_new(
+      answer,
+      "error",
+      janssonCheckPtr(json_string("Could not acquire read lock in one second."))));
+  } else {
+    // Fill JSON
+    auto &manager = ConfigurationManager::getManager(lock);
+
+    {
+      // Topology
+      json_auto_t *topo = janssonCheckPtr(json_object());
+      const char * const mode =
+        manager.getTopologySource() == ConfigurationManager::Source::DB ? "Database" : "XML";
+      janssonCheck(
+        json_object_set_new(topo, "mode", janssonCheckPtr(json_string(mode))));
+
+      json_auto_t *array = janssonCheckPtr(json_array());
+      for (const auto &source : manager.getTopologySourceDetails()) {
+        janssonCheck(json_array_append_new(array, json_string(source.data())));
+      }
+      janssonCheck(json_object_set(topo, "sources", array));
+
+      janssonCheck(json_object_set(answer, "topology", topo));
+    }
+
+    {
+      // Objects
+      json_auto_t *obj = janssonCheckPtr(json_object());
+
+      const char * const mode =
+        manager.getObjectSource() == ConfigurationManager::Source::DB ? "Database" : "XML";
+      janssonCheck(
+        json_object_set_new(obj, "mode", janssonCheckPtr(json_string(mode))));
+
+      json_auto_t *array = janssonCheckPtr(json_array());
+      for (const auto &source : manager.getObjectSourceDetails()) {
+        janssonCheck(json_array_append_new(array, json_string(source.data())));
+      }
+      janssonCheck(json_object_set(obj, "sources", array));
+
+      janssonCheck(json_object_set(answer, "objects", obj));
+    }
+
+    {
+      // Statistics
+      json_auto_t *statsJson = janssonCheckPtr(json_object());
+      auto stats = manager.getStatistics();
+
+      janssonCheck(json_object_set_new(
+        statsJson,
+        "amc13-count",
+        janssonCheckPtr(json_integer(stats.amc13Count))));
+      janssonCheck(json_object_set_new(
+        statsJson,
+        "amc-count",
+        janssonCheckPtr(json_integer(stats.amcCount))));
+      janssonCheck(json_object_set_new(
+        statsJson,
+        "oh-count",
+        janssonCheckPtr(json_integer(stats.ohCount))));
+      janssonCheck(json_object_set_new(
+        statsJson,
+        "vfat-count",
+        janssonCheckPtr(json_integer(stats.vfatCount))));
+      janssonCheck(json_object_set(answer, "statistics", statsJson));
+    }
+  }
+
+  // Write to stream
+  char *encoded = json_dumps(answer, 0);
+  *out << encoded;
+  free(encoded);
```

Comment:
This `free` could indeed be avoided using a smart pointer, but it may be overkill.